src/Oscoin/Storage/Block/Orphanage.hs

```diff
@@ -270,17 +297,99 @@ storeOrphan
     -> Orphanage c tx s
 storeOrphan b o = o { orphans = Map.insert (blockHash b) b (orphans o) }
 
+{------------------------------------------------------------------------------
+
+                                Pruning chains
+
+ If a suitable 'ChainCandidate' has been found, the winning chain needs to be
+ removed from the 'Orphanage'. We offer two different functions, depending from
+ the context: 'pruneOrphanageShallow' removes only the winning chain, but it
+ doesn't touch any chain branching from any of the blocks of the chain being
+ pruned. If that's not the desired behaviour, you might want to consider using
+ 'pruneOrphanageDeep'. In pictures:
+
+
+          ORIGINAL                   SHALLOW                         DEEP
+
+                 |                         |                            |
+                 |                         |                            |
+                 |                         |                            |
+    Winning      |                         |                            |
+    Candidate    |                         |                            |
+            +-+  |                         |                            |
+            | | /\                         \                            \
+            | |-  -\                        -\                           -\
+            |<|     >                         >                            >
+            |||
+            |||
+            |v|
+            |-|                       /-
+          /-|||                     /-
+        /-  |||                   /-
+      <-    |||                 <-
+            |v|                         \
+            | |\                         \
+            | | v                         v
+            | |
+            +-+
+
+------------------------------------------------------------------------------}
+
+-- | Like 'pruneOrphanageShallow', but also prunes any other candidate chain
+-- branching off from /any/ block of the chain being pruned.
+pruneOrphanageDeep
+    :: (Ord (BlockHash c)
+       , Eq s
+       )
+    => BlockHash c
+    -> ChainCandidate c s
+    -> Orphanage c tx s
+    -> Orphanage c tx s
+pruneOrphanageDeep rootHash bestFork@ChainCandidate{candidateChain} o =
+    foldl' (flip bulkDelete)
+           (pruneOrphanageShallow rootHash bestFork o)
+           (toList candidateChain)
+
 -- | Given a candidate which won, delete the orphans from the store and
--- the chain from the candidates
-pruneOrphanage
-    :: Ord (BlockHash c)
+-- the chain from the candidates.
+pruneOrphanageShallow
+    :: forall c tx s.
+       ( Ord (BlockHash c)
+       , Eq s
+       )
     => BlockHash c
```

Comment:
I’m struggling to understand the significance of `BlockHash`. In my understanding a precondition is that `rootHash` is the parent hash of the `ChainCandidate`. If so we should definitely document that.

Reply:
yes, your understanding is correct -- we could in theory derive it directly from the `ChainCandidate`, at the cost of calling `lookupChainRoot`. The advantage is that it would simplify the API, making things easier to use for end-users. Will give it a shot.

Reply:
Or you could add it as a field to `ChainCandidate`

Reply:
34492b94